dependencies-mindset-background.Rmd

```diff
@@ -0,0 +1,598 @@
+
+These are basically regarded as free dependencies and can be added to `DESCRIPTION` via `usethis::use_tidy_dependencies()` (which also does a few more things).
+It should come as no surprise that these packages have a very small dependency footprint.
+
+```{r}
+tools::package_dependencies(c("rlang", "cli", "glue", "withr", "lifecycle"))
+```
+
+::: callout-warning
+## Submitting to CRAN
+
+Under certain configurations, including those used for incoming CRAN submissions, `R CMD check` issues a `NOTE` if there are 20 or more "non-default" packages in `Imports`:
+
+    N  checking package dependencies (1.5s)
+       Imports includes 29 non-default packages.
+       Importing from so many packages makes the package vulnerable to any of
+       them becoming unavailable.  Move as many as possible to Suggests and
+       use conditionally.
+
+Our best advice is to try hard to comply, as it should be rather rare to need so many dependencies and it's best to eliminate any `NOTE` that you can.
+Of course, there are exceptions to every rule and perhaps your package is one them.
+In that case, you may need to argue your case.
+It is certainly true that many CRAN packages violate this threshold.
+:::
+
+### Whether to Import or Suggest {#sec-dependencies-imports-vs-suggests}
+
+The [withr package](https://withr.r-lib.org) is a good case study for deciding whether to list a dependency in `Imports` or `Suggests`.
+Withr is very useful for writing tests that clean up after themselves.
+Such usage is compatible with listing withr in `Suggests`, since regular users don't need to run the tests.
+But sometimes a package might also use withr in its own functions, perhaps to offer its own `with_*()` and `local_*()` functions.
+In that case, withr should be listed in `Imports`.
+
+`Imports` and `Suggests` differ in the strength and nature of dependency:
+
+-   `Imports`: packages listed here *must* be present for your package to work.
+    Any time your package is installed, those packages will also be installed, if not already present.
+    `devtools::load_all()` also checks that all packages in `Imports` are installed.
+
+    It's worth pointing out that adding a package to `Imports` ensures it will be installed and **that is all it does**.
+    It has nothing to do with actually importing functions from that package.
+    See @sec-dependencies-in-imports for more about how to use a package in `Imports`.
+
+-   `Suggests`: your package can use these packages, but doesn't require them.
+    You might use suggested packages for example datasets, to run tests, build vignettes, or maybe there's only one function that needs the package.
+
+    Packages listed in `Suggests` are not automatically installed along with your package.
+    This means that you can't assume the package is available unconditionally.
+    See @sec-dependencies-in-suggests for how to check whether a suggested package is installed.
```

Comment:
But you can assume developers have them.

Reply:
added